api/src/main/java/net/kyori/adventure/text/serializer/ComponentSerializer.java

```diff
@@ -137,4 +138,39 @@
 
     return this.serialize(component);
   }
+
+  /**
+   * A {@link Component} serializer and deserializer builder.
+   *
+   * @param <I> the input component type
+   * @param <O> the output component type
+   * @param <R> the serialized type
+   *
+   * @since 4.14.0
+   */
+  interface Builder<I extends Component, O extends Component, R> {
+
+    UnaryOperator<String> DEFAULT_NO_OP = UnaryOperator.identity();
+    UnaryOperator<Component> DEFAULT_COMPACTING_METHOD = Component::compact;
+
+    /**
+     * Specify a function that takes the component at the end of the parser process.
+     * <p>By default, this compacts the resulting component with {@link Component#compact()}.</p>
+     *
+     * @param postProcessor method run at the end of parsing
+     * @return this builder
+     * @since 4.14.0
+     */
+    @NotNull Builder<I, O, R> postProcessor(final @NotNull UnaryOperator<Component> postProcessor);
+
+    /**
+     * Specify a function that takes the string at the start of the parser process.
+     * <p>By default, this does absolutely nothing.</p>
+     *
+     * @param preProcessor method run at the start of parsing
+     * @return this builder
+     * @since 4.14.0
+     */
+    @NotNull Builder<I, O, R> preProcessor(final @NotNull UnaryOperator<String> preProcessor);
```

Comment:
Should be `UnaryOperator<R>` instead, as the result of deserialization might be different than `String`.